### ISSUE TITLE
Default DisplayNameResolver cannot handle classes derived from DisplayAttribute or DisplayNameAttribute

### ISSUE THREAD

**Issue by fsateler** [JeremySkinner/FluentValidation#286]
The default resolver tests against the class name explicitly:

https://github.com/JeremySkinner/FluentValidation/blob/11e46cba4727f2e7e0a06de51b4d0e4bd92ba341/src/FluentValidation/ValidatorOptions.cs#L76-L92

This was apparently done because referencing DataAnnotations caused trouble for some people (https://github.com/JeremySkinner/FluentValidation/commit/344d6a4621e32c7005b35bb8e5c8fe63630a212a).  This commit unfortunately does not mention what is the issue.

This has the problem of not allowing derived classes (as they will very likely use a different name), as in issue #192 .

It would be great if derived classes could be used as-is.

**Comment by JeremySkinner** [comment]
There was a .net framework bug that caused weird exceptions if FV referenced DataAnnotations directly. I can't remember the specific details of what the exception was, but this was the workaround. 

I seem to remember this was a problem under net35, so it may no longer be an issue as we don't support net35 anymore, and we can get rid of the workaround 

I'll try and do some investigation into this, but probably won't get to it for a while I'm afraid.

**Comment by RedTaz** [comment]
I had to dig in to this today.

Since the workaround in `ValidatorOptions` is doing basic string comparison of the type names it should probably recursively check the base type names of each of the custom attributes too.

We're having to circumvent the workaround (a workaround for the workaround?) by redefining the `DisplayNameResolver` within our application. The code below is taken from an old commit of `ValidatorOptions`

```
            ValidatorOptions.DisplayNameResolver = (type, memberInfo, expression) => {

                if (memberInfo == null) return null;
                string name = null;

                var displayAttribute = (DisplayAttribute)Attribute.GetCustomAttribute(memberInfo, typeof(DisplayAttribute));

                if (displayAttribute != null) {
                    name = displayAttribute.GetName();
                }

                if (string.IsNullOrEmpty(name)) {
                    // Couldn't find a name from a DisplayAttribute. Try DisplayNameAttribute instead.
                    var displayNameAttribute = (DisplayNameAttribute)Attribute.GetCustomAttribute(memberInfo, typeof(DisplayNameAttribute));
                    if (displayNameAttribute != null) {
                        name = displayNameAttribute.DisplayName;
                    }
                }

                return name;

            };
```

**Comment by JeremySkinner** [comment]
I've committed the change to use the attribute references directly for version 7.5 if you're running on netstandard 2.0 or net45. For netstandard1.0, it'll continue to use the original behaviour.

**Comment by JeremySkinner** [comment]
I've released 7.5 to nuget today that includes this fix.
